template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/BlackListedLoadBasedAutoscaleRole.java

```diff
@@ -0,0 +1,12 @@
+package com.sequenceiq.cloudbreak.cmtemplate;
+
+public enum BlackListedLoadBasedAutoscaleRole {
```

Comment:
Using a blacklist will need all services from CM to be added in here; or make a new release of CB when we add new services.
If I'm not mistaken, with this set of services - the OperationDB (HBase) clusters will qualify for both Time and Load based AutoScaling.

A Whitelist is an alternate, which comes with its own set of problems. i.e. if "NODEMANAGER" is the only role there, the GATEWAY roles that go along with the NODEMANAGER in a hostgroup will cause trouble.

Maybe the GATEWAYs just need to be removed from processing. i.e. A GATEWAY (unless it is special) does not block a whitelisted role from being scaled, or in case of BLACKLISTING - GATEWAY roles are just filtered out. If that leads to no roles left for a hostGroup - the hostGroup is not scalable (but maybe we should allow this to be time scalable).

I had a slightly different take on how this would work - based more on a 'Role' -> Capability mechanism - with all roles on a hostGroup combined together. A separate enum per capability (this patch) works as well.

Reply:
### Regarding OpDB

[Gateway, Master, Leader, Worker] of these,

Time based -> [Gateway, Leader] # Master has NAMENODE and Worker has DATANODE
Load based -> [] # Gateway and Leader has GATEWAY

### General problem

I was thinking of Blacklist filtering, followed by forceful whitelisting. Which is kind of acheived now using the fallback filter, just that it is counter-intuitive. 

We have two fields in the CM template, roleType, refName. We see the problems with roleType. This is existing implementation for Nifi cluster, which works. However when force fitted into other templates, it struggles. RefName is interesting and what you asked for in the previous comment. NODEMANAGER for worker gets yarn-NODEMANAGER-WORKER and compute gets yarn-NODEMANAGER-COMPUTE. I am not sure if we can rely on this naming, and how it will change for custom templates.

Reply:
roleType is what I was referring to. Service.RoleType fully qualifies an entity.
YARN.NODEMANAGER
YARN.GATEWAY
HDFS.GATEWAY

Yup, OpDB will be handled because of the presence of the DataNode. Think it is easier to reason with 'Scaling supported only for NodeManager'.

IAC, this works. The API for the UI, and whatever will be added for Periscope, remains unchanged irrespective of blacklist vs whitelist.

Any thoughts on ignoring GATEWAYs in the processing?

Also, do we want to allow schedule policies to be defined for GATEWAY nodes? (What if Hue is running on one of these nodes - is that safe?. Should Hue be in the blacklist?)

Reply:
> Service.RoleType fully qualifies an entity.
> YARN.NODEMANAGER

For me it did not. I want to whitelist YARN.NODEMANAGER-COMPUTE. Here compute is placement. If I whitelist YARN.NODEMANAGER then worker is also whitelisted. We can over come that with first blacklist and then whitelist. Remove the ones with DATANODE then add the ones with NODEMANAGER.

I chose DATANODE over NODEMANAGER for this reason.

> Any thoughts on ignoring GATEWAYs in the processing?

Yes, it depends on GATEWAY node support.

We can ignore GATEWAY nodes if it simplifies things. Hue is not a blocker, it can scale well. The Knox topology does not change. Hue sits behind CM controlled load-balancer. DAS, and Livy could be added as blacklists. Since we have not thought through this, we can ignore it for now.

Reply:
YARN.NODEMANAGER-COMPUTE is a dynamic name, and can be renamed to anything in a custom template. Think you mentioned that further up as well. YARN.NODEMANAGER, however, is determined by CM.

We are not going to be testing gateway node auto-scaling. Also, I don't think we're allowing more than one autoScaling policy for a cluster at the moment. Think it is better to make sure the GATEWAY is not auto-scalable for the moment. There's also limited value in this, since the hostnames will change each time a scale operation is run.

So - we're going with a BLACKLIST for now, and trying to put enough in there to ensure that only the compute group can be automatically scaled.

Reply:
Does it make sense to disable 'gateway' hosts from AutoScaling?, since we are not testing them.

Reply:
Yep, I will remove both gateway role and gateway node group.